feat: make it into a DAG-CBOR library

vmx opened the pull request:
This library is not a general purpose CBOR library anymore, but
specifically targeting DAG-CBOR. Use `Ipld` as value type.

Other changes:

 - Sort the map keys as DAG-CBOR requires it
 - Remove support for unit structs
 - Floats are always serialized as 64-bit values
 - CBOR null and undefned are deserialized into `None` and not
   into unit (`()`).

vmx committed:
feat: make it into a DAG-CBOR library

This library is not a general purpose CBOR library anymore, but
specifically targeting DAG-CBOR. Use `Ipld` as value type.

Other changes:

 - Sort the map keys as DAG-CBOR requires it
 - Remove support for unit structs
 - Floats are always serialized as 64-bit values
 - CBOR null and undefned are deserialized into `None` and not
   into unit (`()`).

vmx committed:
Revert "Merge pull request #172 from rklaehn/tags"

We don't need full tag support. We only need support for tag 42 (CID). Hence
remove the tag support and add support for tag 42 in a subsequent commit.

This reverts commit 16c3c1c76c6c79e94f7ac6c3b2a6bf8d07c09e9c, reversing
changes made to dbcc31f00bc2e01152b51d5af1a10ed860b7b829.

Stebalien commented on src/de.rs:
```diff
@@ -725,34 +713,24 @@ where
-                let tag = u64::from(byte) - 0xc0;
-                self.handle_tagged_value(tag, visitor)
-            }
+            // Only tag 42 is supported, hence we refuse parsing any tags here.
```
any _other_ tags.

Stebalien commented on src/de.rs:
```diff
@@ -1358,3 +1344,78 @@ where
+    #[inline]
+    fn deserialize_bytes<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
+        // Match on the major type, it must be a byte string (major type 2)
+        let len = match self.0.parse_u8()? {
```
Can we not reuse our vec deserialization code?

vmx replied:
What do you mean with "vec deserialization code"?

Stebalien replied:
I mean, we already have logic for deserializing byte buffers, is there any way to reuse that here? (not critical, just a thought)

vmx replied:
Oh I see. yes we have, but it's not really worth it (given the current code structure of this library).

Stebalien commented on src/de.rs:
```diff
@@ -1358,3 +1344,78 @@ where
+    }
+
+    #[inline]
+    fn deserialize_bytes<V: de::Visitor<'de>>(self, visitor: V) -> Result<V::Value> {
```
If I try to parse IPLD into a `json::Value`, will this not allow us to parse into a vec? Can we have `deserialize_newtype_struct be the only entry point?

vmx replied:
I don't see the problem of non-DAG codecs to deserialize things to whatever.

Stebalien commented on src/de.rs:
```diff
@@ -1358,3 +1344,78 @@ where
+{
+    type Error = Error;
+
+    fn deserialize_any<V: de::Visitor<'de>>(self, _visitor: V) -> Result<V::Value> {
```
Attempting to decode a CID into something _else_ shouldn't panic. I.e.:

1. User calls `deserialize_any` on the root deserializer.
2. Our deserializer calls `visit_newtype_struct` with this deserializer.
3. User tries to visit the wrong type (because they're not expecting a CID.

Stebalien commented on src/ser.rs:
```diff
@@ -741,3 +762,151 @@ where
+    type SerializeStructVariant = ser::Impossible<Self::Ok, Self::Error>;
+
+    fn serialize_bool(self, _value: bool) -> Result<Self::Ok> {
+        unreachable!();
```
So, this should _actually_ be unreachable, but it would be nice to have an error anyways.

vmx committed:
fix: CID (de)serializer: error instead of panic

vmx committed:
chore: fix comment

vmx committed:
fix: make it at least compile with alloc feature only

You can now compile it with only the alloc feature being enabled, it
doesn't need std.

Stebalien commented on Cargo.toml:
```diff
@@ -30,4 +33,6 @@ default = ["std"]
-tags = []
+
+[patch.crates-io]
+libipld-core = { git = "https://github.com/ipld/libipld", branch = "ipld-serde" }
```
update?

Stebalien commented on src/de.rs:
```diff
@@ -725,34 +715,24 @@ where
-            0xf6 => visitor.visit_unit(),
-            0xf7 => visitor.visit_unit(),
+            0xf6 => visitor.visit_none(),
+            0xf7 => visitor.visit_none(),
```
We should reject undefined, right? That's what cbor-gen does (and that's what the dag-cbor spec says).

vmx replied:
I think there are still more things that could be stricter. Focus was on "getting things work". But that's an easy one to change.

Stebalien replied:
We can pick this up later.

Stebalien replied:
Ah, I see you already did.

Stebalien commented on src/de.rs:
```diff
@@ -1358,3 +1346,78 @@ where
+            _ => Err(self.0.error(ErrorCode::UnexpectedCode))?,
+        };
+
+        match self.0.read.read(len)? {
```
what if length is 0? I believe the code below will panic.

vmx replied:
Good catch!

Stebalien reviewed (approved):
Two small issues, otherwise lgtm.

vmx committed:
chore: patching libipld-core is no longer needed

vmx committed:
fix: reject undefined

DAG-CBOR doesn't support `undefined`.

vmx committed:
fix: don't panic on zero sized CID

Even if passing in an invalid CID, there shouldn't be a panic. If
there is a zero sized CID, it will no error.
